sys/net/gnrc/transport_layer/tcp/gnrc_tcp_fsm.c

```diff
@@ -307,9 +307,11 @@ static int _fsm_call_recv(gnrc_tcp_tcb_t *tcb, void *buf, size_t len)
     /* Read data into 'buf' up to 'len' bytes from receive buffer */
     size_t rcvd = ringbuffer_get(&(tcb->rcv_buf), buf, len);
 
-    /* If receive buffer can store more than GNRC_TCP_MSS: open window to available buffer size */
-    if (ringbuffer_get_free(&tcb->rcv_buf) >= GNRC_TCP_MSS) {
-        tcb->rcv_wnd = ringbuffer_get_free(&(tcb->rcv_buf));
+    /* Reopen window if recv buffer can hold a MSS sized payload and FIN was not received. */
+    uint16_t buf_free = ringbuffer_get_free(&tcb->rcv_buf);
+
+    if (buf_free >= GNRC_TCP_MSS && tcb->state != FSM_STATE_CLOSE_WAIT) {
+        tcb->rcv_wnd = buf_free;
```

Comment:
This change seems to be unrelated (i.e. it should at least go into a separate commit) and I'm not sure if it is valid. Is it guaranteed that `tcb->rcv_buf`does not change between the `if` and the assignment to `tcb->rcv_wnd`? Or asked differently: is the calling thread to `ringbuffer_get_free()` the same that writes to the ring buffer?

Reply:
The calling threads are different, but all calls to the FSM are synchronized via a mutex. 
There is no problem there.

Reply:
Separate PR: #12379